File: src/net/zomis/tttultimate/TTPlayer.java
package net.zomis.tttultimate;

public enum TTPlayer {

	NONE, X, O, XO;
	
	public TTPlayer next() {
		if (!isRealPlayer())
			throw new UnsupportedOperationException("Only possible to call .next() on a real player but it was called on " + this);
		return this == X ? O : X;
	}
	public TTPlayer and(TTPlayer other) {
		if (this == NONE || other == NONE || other == null)
			return NONE;
		if (isRealPlayer())
			return this == other ? this : NONE;
		return other;
	}
	public boolean isRealPlayer() {
		return this == X || this == O;
	}
	public static boolean isRealPlayer(TTPlayer winner) {
		return winner != null && winner.isRealPlayer();
	}
	public TTPlayer or(TTPlayer other) {
		if (this == NONE || other == null)
			return other;
		if (other == NONE)
			return this;
		if (this == XO)
			return this;
		if (this != other)
			return XO;
		return this;
	}
	
}

File: uttt-common/src/main/kotlin/net/zomis/tttultimate/TTPlayer.kt
package net.zomis.tttultimate

enum class TTPlayer {

    /**
     * No player
     */
    NONE,
    X,
    O,
    /**
     * Both players has succeeded
     */
    XO,
    /**
     * Neither player (both have tried, none has succeeded)
     */
    BLOCKED,
    ;

    val isExactlyOnePlayer: Boolean
        get() = this == X || this == O

    operator fun next(): TTPlayer {
        if (!isExactlyOnePlayer) {
            throw UnsupportedOperationException("Only possible to call .next() on a real player but it was called on " + this)
        }
        return if (this == X) O else X
    }

    /**
     * Determine if this player is (also) another player.<br></br>
     * This is the same as `this.and(other) == other`
     *
     * @param other
     * @return
     */
    fun `is`(other: TTPlayer): Boolean {
        return this.and(other) == other
    }

    fun and(other: TTPlayer): TTPlayer {
        if (this == NONE || other == NONE) {
            return NONE
        }
        if (isExactlyOnePlayer && other.isExactlyOnePlayer) {
            return if (this == other) this else NONE
        }
        if (this == BLOCKED || other == BLOCKED) {
            return other
        }
        return if (this == XO) other else other.and(this)
    }

    fun or(other: TTPlayer): TTPlayer {
        if (this == NONE) {
            return other
        }
        if (other == NONE) {
            return this
        }
        if (this == XO) {
            return this
        }
        if (this == BLOCKED || other == BLOCKED) {
            return BLOCKED
        }
        return if (this != other) XO else this
    }

    companion object {
        fun isExactlyOnePlayer(winner: TTPlayer?): Boolean {
            return winner != null && winner.isExactlyOnePlayer
        }
    }

}
